### PR TITLE
SPG Trainer

### PR DESCRIPTION
**Pull request by nicholas-voigt**
Improved Permutation Model Training with the following changes:
- improved numerical stability by transitioning to log-probabilities in soft permutation matrix
- changed critic to be solely dependent on state and not state and action
- added critic warm-up phase
- changed actor loss function to Reinforce with critic baseline, loss is relative advantage over baseline.
- critic loss is MSE loss between critic estimation and actual tour cost
- updated evaluation function for inference

### CODE REVIEW

File: model/components.py [hunk 1]
```diff
@@ -354,24 +354,22 @@ def sinkhorn(self, scores: torch.Tensor) -> torch.Tensor:
         # pre-normalize for numerical stability
         scores = scores - scores.max(dim=-1, keepdim=True)[0]
         for _ in range(self.gs_iters):
-            scores = scores - torch.logsumexp(scores, dim=2, keepdim=True)  # row norm
-            scores = scores - torch.logsumexp(scores, dim=1, keepdim=True)  # col norm
-        return torch.softmax(scores, dim=-1)
+            scores = F.log_softmax(scores, dim=2)  # row norm
+            scores = F.log_softmax(scores, dim=1)  # col norm
```
[the remark this answers is not in the document]
Comment (nicholas-voigt) [reply model/components.py:358]: In my understanding log_softmax(scores) is equivalent to scores - logsumexp(scores). And the alternating normalization of columns and rows pushes the matrix close to a doubly-stochastic matrix in log-space.
